spec/ruby/language/predefined/toplevel_binding_spec.rb

```diff
@@ -0,0 +1,24 @@
+require File.expand_path('../../../spec_helper', __FILE__)
+
+describe "The TOPLEVEL_BINDING constant" do
+  before :each do
+    @binding_toplevel_variables = eval(ruby_exe(fixture(__FILE__, "toplevel_binding_variables.rb")))
+    @binding_toplevel_id = eval(ruby_exe(fixture(__FILE__, "toplevel_binding_id.rb")))
+  end
+  
+  it "includes local variables defined in the same file" do
+    @binding_toplevel_variables.should include(:foo)
+  end
+
+  it "does not include local variables defined in a required file" do
+    @binding_toplevel_variables.should_not include(:bar)
+  end
+
+  it "does not include local variables defined in eval" do
+    @binding_toplevel_variables.should_not include(:baz)
+  end
+
+  it "is always the same object for all top levels" do
+    @binding_toplevel_id.uniq.size.should == 1
```

Comment:
Any idea how to fix that one or what MRI/JRuby do?

Reply:
I think it's possibly a single frame for all files, but cleared of local variables for each new file. That's really tricky for us - we don't control creation of the frame!

Reply:
Maybe we should create a single global `TOPLEVEL_BINDING` frame, and then in top level scopes we always assign into it as an outer scope?

Reply:
That would not capture the locals of the main script though, isn't it?
I think we need to create the TOPLEVEL_BINDING early, and then patch its frame when just before executing the main script.